subsys/net/ip/tcp2.c

```diff
@@ -413,6 +417,11 @@ static void tcp_send_process(struct k_work *work)
 	if (conn && conn->in_retransmission) {
 		k_delayed_work_submit(&conn->send_timer, K_MSEC(tcp_rto));
 	}
+
+out:
+	if (conn) {
```

Comment:
I don't like these `if (conn)` checks in this function. The only way they will be true is when `conn = NULL;` is set in this function. I think that it is simpler to just exit early instead of setting `conn = NULL;` and then checking `if (conn)` twice.

Reply:
I suppose we could return in line 394 instead of setting `conn=NULL`. I think the idea here has been to have only one exit point in this function so we have the out: label at the end. I would leave this as is.

Reply:
> I think the idea here has been to have only one exit point in this function so we have the out: label at the end.

+1. This is common "exception handling" pattern in C (including pattern of label naming). Wouldn't make sense to (mis)"optimize" it for this particular case, risking that it may be problematic in case of further changes.

Reply:
So maybe in case of `conn = NULL;` lets jump after all `if (conn)` statements, as there is no reason to go through them? Such goto would be of course equivalent of `return;`, but technically there would be still one exit point.

Another idea would be to add `bool unref = false;`, which would be set `unref = true` instead of `conn = NULL`. Something like:
```
static void tcp_send_process(struct k_work *work)
{
	struct tcp *conn = CONTAINER_OF(work, struct tcp, send_timer);
	struct net_pkt *pkt;
	bool unref = false;

	k_mutex_lock(&conn->lock, K_FOREVER);

	...

	if (conn->in_retransmission) {
		if (conn->send_retries > 0) {
			...
		} else {
			unref = true;
			goto out;
		}
	} else {
		...
	}

	if (conn->in_retransmission) {
		k_delayed_work_submit(&conn->send_timer, K_MSEC(tcp_rto));
	}

out:
	k_mutex_unlock(&conn->lock);

	if (unref) {
		tcp_conn_unref(conn);
	}
}
```

Regarding label naming https://www.kernel.org/doc/html/latest/process/coding-style.html#centralized-exiting-of-functions documents how good label names should look like. Such naming helps if there are more labels, as each one describes what it does. `out` isn't particularly good name to say "unlock packet and return".

Reply:
> Regarding label naming https://www.kernel.org/doc/html/latest/process/coding-style.html#centralized-exiting-of-functions documents how good label names should look like. Such naming helps if there are more labels,

Right, it helps in case there're many labels, and in case there's one, it helps if it's named consistently, showing it's general purpose. And this purpose is effectively like "finally" clause in languages which support explicit exception handling. I won't argue here that "finally" would make a good label name, but if we have a pattern of using "out" for this purpose, sticking to it makes sense IMHO.

Reply:
> Another idea would be to add `bool unref = false;`, which would be set `unref = true` instead of `conn = NULL`.

Sure, this looks good. I will send a new version.